Earthfile

```diff
@@ -38,257 +60,379 @@ setup:
+
+###############################################################################
+### Image Stage
+# This stage is responsible for creating the final image with the ROS 2
+# workspace. The image is saved only if the build-test stage is successful.
+###############################################################################
+prepare-image:
+  FROM +pre-installation
+
+  # Add missing dependencies
+  RUN apt update \
+    && apt install -y libspdlog-dev \
+      python3-numpy \
+      tzdata \
+      sudo \
+      ros-dev-tools
+  RUN pip3 install pyyaml \
+      lark \
+      packaging \
+      netifaces \
+      catkin_pkg \
+      psutil --break-system-packages
+
+  # Prepare the image
+  RUN mkdir -p ${SPACEROS_DIR}
+  COPY +rosdep/rosdeps.sh ${SPACEROS_DIR}/rosdeps.sh
+  RUN bash ${SPACEROS_DIR}/rosdeps.sh
+
+###############################################################################
+### Prepare Image Stage
+# This stage is responsible for preparing the image with the ROS 2 workspace.
+# The image is saved only if the build-test stage is successful.
+###############################################################################
 image:
-  FROM +rosdep
-  ARG VCS_REF
-  ARG VERSION='jazzy'
+  ARG --required IMAGE_VARIANT
+  FROM +prepare-image
+
+  COPY +build/spaceros ${SPACEROS_DIR}
+  COPY +sources/exact.repos ${SPACEROS_DIR}/scripts/spaceros.repos
+  COPY scripts/generate-repos.sh scripts/merge-repos.py ${SPACEROS_DIR}/scripts/
+  RUN chmod +x ${SPACEROS_DIR}/scripts/generate-repos.sh ${SPACEROS_DIR}/scripts/merge-repos.py \
+    && mv ${SPACEROS_DIR}/rosdeps.sh ${SPACEROS_DIR}/scripts/rosdeps.sh
+
+  # Post Installation
+  DO +POST_INSTALLATION --IMAGE_VARIANT=${IMAGE_VARIANT}
+
+  # Clear Apt and Pip cache
+  RUN rm -rf /var/lib/apt/lists/* /tmp/pip-reqs /var/cache/apt/archives \
+    && apt-get clean \
+    && pip cache purge
+
+  # Add user and group
+  RUN useradd -m -s /bin/bash ${USERNAME} \
+    && chown -R ${USERNAME}:${USERNAME} ${SPACEROS_DIR}/scripts \
+    && echo "${USERNAME} ALL=(ALL) NOPASSWD:ALL" > /etc/sudoers.d/${USERNAME}
+  RUN echo "source ${SPACEROS_DIR}/setup.bash" >> ${HOME}/.bashrc
+
+  RUN chown -R ${USERNAME}:${USERNAME} ${HOME}
+
+  USER ${USERNAME}
+  WORKDIR ${HOME}
+
+  # Only save image if build-test is successful
+  IF [ "${SKIP_BUILD_TEST}" = "true" ]
+    SAVE IMAGE ${IMAGE_NAME}:${IMAGE_VARIANT}
+  ELSE
+    BUILD +build-test
+    SAVE IMAGE ${IMAGE_NAME}:${IMAGE_VARIANT}
+  END
+
+###############################################################################
+### Post Installation Stage
+# This stage is responsible for cleaning up the workspace and installing
+# additional dependencies based on the image variant.
+###############################################################################
+POST_INSTALLATION:
+  FUNCTION
+  ARG --required IMAGE_VARIANT
+
+  # Remove workspace if image variant is core or dev
+  RUN rm -rf ${WORKSPACE_DIR}
```

Comment:
Thinking about this, and tagging you @Bckempa since we were talking about it, do we want the workspace to be preserved in the "dev" image so that we can just build and test without recloning anything?

Reply:
I think we could have the workspace so we have some common practices between main and dev images.